### PR TITLE
feat(sdk-crash-detection): Call code in post processing

### LINKED ISSUE #46175: Call code in post processing
Identify the proper location to call the code of #44342.

### Option 1: Event Manager
The `event_manager` similar to where we call `detect_performance_problems`. 

https://github.com/getsentry/sentry/blob/fc1d72ee868af1103f2751e2fa0f1928d3d2ba4c/src/sentry/event_manager.py#L2491

### Option 2:  Issue platform detectors

The purpose of [detectors](https://github.com/getsentry/sentry/tree/fc1d72ee868af1103f2751e2fa0f1928d3d2ba4c/src/sentry/utils/performance_issues/detectors) is to find problems in the customer's dataset, such as N+1 queries, file I/O on the main thread, slow DB query, etc. As #44342 aims to find problems of the SDKs, detectors don't seem like the correct place.

### Option 3: Post Processing
[post_processing](https://github.com/getsentry/sentry/blob/e39eeadd8b67c4c2c11f02d6ce8f2a2ef99bc018/src/sentry/tasks/post_process.py#L1): does some other non-critical event processing. We could add it to the post processing pipeline: https://github.com/getsentry/sentry/blob/fc1d72ee868af1103f2751e2fa0f1928d3d2ba4c/src/sentry/tasks/post_process.py#L976-L992

### PR DESCRIPTION
**Pull request by philipphofmann**
## If you are not @getsentry/ingest you can skip this PR.

Merging master into feature branches accidentally requested a review from almost all teams at Sentry. I'm sorry about this.

## PR Details

Call SDK crash detection code in post-processing. I had to adapt the code a bit, cause I tested it with a dictionary instead of a stored event. I adapted the tests to use the same approach as the post-processing tests. The logic of the tests didn't change. I refactored all of them to work with stored events. 

The next step is to store the SDK crash event to a specific sentry project with https://github.com/getsentry/sentry/issues/46177

I wasn't able to get parameterized tests working with test mixin approach. Does anybody know if that's possible?

Fixes GH-46175

### CODE REVIEW

**Review by jjbayer — CHANGES_REQUESTED** [review]
> I wasn't able to get parameterized tests working with test mixin approach. Does anybody know if that's possible?

`pytest.mark.parametrize` does not work in combination with `unittest.TestCase`, unfortunately: https://github.com/pytest-dev/pytest/issues/541

Why are you moving from standalone (pytest) functions to class-based tests, if I may ask?

File: src/sentry/utils/sdk_crashes/sdk_crash_detection.py [hunk 1]
```diff
@@ -27,23 +29,40 @@ def __init__(
...
 
     def detect_sdk_crash(self, event: Event) -> None:
-        if event.get("type", None) != "error" or event.get("platform") != "cocoa":
+
+        should_detect_sdk_crash = (
+            event.group
+            and event.group.issue_category == GroupCategory.ERROR
+            and event.group.platform == "cocoa"
+        )
+        if should_detect_sdk_crash is False:
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/sdk_crash_detection.py:38]: Uh, nice. I am still getting used to Python. Thanks 👍

File: src/sentry/utils/sdk_crashes/sdk_crash_detection.py [hunk 2]
```diff
@@ -27,23 +29,40 @@ def __init__(
...
+                sdk_crash_event.data, "contexts", "sdk_crash_detection", value={"detected": True}
+            )
             self.sdk_crash_reporter.report(sdk_crash_event)
+
+
+_crash_reporter = SDKCrashReporter()
+_cocoa_sdk_crash_detector = CocoaSDKCrashDetector()
+_event_stripper = EventStripper(sdk_crash_detector=_cocoa_sdk_crash_detector)
+
+sdk_crash_detection = SDKCrashDetection(_crash_reporter, _cocoa_sdk_crash_detector, _event_stripper)
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/sdk_crash_detection.py:68]: Yes, that makes sense. Old object-oriented habits. Thanks for pointing it out. You don't need a class for everything.

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 3]
```diff
@@ -27,22 +27,25 @@ def __init__(
...
             stripped_frames = self._strip_frames(frames)
-            new_event["exception"]["values"][0]["stacktrace"]["frames"] = stripped_frames
+            new_event_data["exception"]["values"][0]["stacktrace"]["frames"] = stripped_frames
 
-        debug_meta_images = get_path(event, "debug_meta", "images")
+        debug_meta_images = get_path(new_event_data, "debug_meta", "images")
         if debug_meta_images is not None:
             stripped_debug_meta_images = self._strip_debug_meta(debug_meta_images, stripped_frames)
-            new_event["debug_meta"]["images"] = stripped_debug_meta_images
+            new_event_data["debug_meta"]["images"] = stripped_debug_meta_images
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/event_stripper.py]: I thought Sentry maybe needed it to display the event properly. If you say everything works fine without the dSYMs, I'm happy to drop them.

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 4]
```diff
@@ -27,22 +27,25 @@ def __init__(
...
 
-        return new_event
+        event.data = new_event_data
+        return event
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/event_stripper.py]: I fully agree. To keep the scope of this PR small, I created another issue to fix this https://github.com/getsentry/sentry/issues/49935 in another PR. Are you okay with that, @jjbayer?

**Comment by philipphofmann** [comment]
> Why are you moving from standalone (pytest) functions to class-based tests, if I may ask?

So I can apply the same approach for creating an event in `test_post_process`. I didn't manage to use the same approach with functions, @jjbayer.

**Comment by jjbayer** [comment]
> > Why are you moving from standalone (pytest) functions to class-based tests, if I may ask?
> 
> So I can apply the same approach for creating an event in `test_post_process`. I didn't manage to use the same approach with functions, @jjbayer.

Yeah so AFAIK you have to chose between using either `@pytest.mark.parametrize` or class-based testing with manual parameterization, whichever you prefer.

**Review by jjbayer — APPROVED** [review]

**Comment by asottile-sentry** [comment]
> > > Why are you moving from standalone (pytest) functions to class-based tests, if I may ask?
> > 
> > 
> > So I can apply the same approach for creating an event in `test_post_process`. I didn't manage to use the same approach with functions, @jjbayer.
> 
> Yeah so AFAIK you have to chose between using either `@pytest.mark.parametrize` or class-based testing with manual parameterization, whichever you prefer.

of note, `TestCase` and class based tests are soft deprecated -- please try and write function tests moving forward (pytest has already begun to remove / unsupport functionality there over time -- the compat layer is only meant as a migration helper)

**Comment by philipphofmann** [comment]
Thanks for pointing that out, @asottile-sentry. I created an issue to change the tests to function tests https://github.com/getsentry/sentry/issues/50061.

**Pull request merged**